### PR TITLE
Fix getValueFromArraybuffer calls to use proper type

### PR DESCRIPTION
**Pull request by Achie72**
Fixes: #225

Signed-off-by: Bela Toth tbela@inf.u-szeged.hu

### CODE REVIEW

File: src/runtime/GlobalObjectBuiltinTypedArray.cpp [hunk 1]
```diff
@@ -456,7 +456,38 @@ Value builtinTypedArrayConstructor(ExecutionState& state, Value thisValue, size_
                 // Repeat, while count > 0
                 while (count > 0) {
                     // Let value be GetValueFromBuffer(srcData, srcByteIndex, srcType).
-                    Value value = srcData->getValueFromBuffer<typename TypeAdaptor::Type>(state, srcByteIndex);
+                    Value value;
+                    switch (val.asObject()->asArrayBufferView()->typedArrayType()) {
```
Comment (rerobika) [inline src/runtime/GlobalObjectBuiltinTypedArray.cpp]: Could you please separate it to an utility function for further reusability?

**Comment by clover2123** [comment]
Would you please add a internal TC related to #225?

**Comment by Achie72** [comment]
@rerobika I think the separation is done, and should work properly.

@clover2123 I added a TC for the original problem, which was caused by improper reading and copying of data inside the buffer objects. I added a few cases testing that as well (copying proper value, handling overflow and signed-unsigned conversions). I hope it's not a problem.

File: src/runtime/ArrayBufferObject.h [hunk 2]
```diff
@@ -98,6 +110,43 @@ class ArrayBufferObject : public Object {
         return true;
     }
 
+    Value getTypedValueFromBuffer(ExecutionState& state, unsigned byteindex, TypedArrayType type)
+    {
+        Value value;
+        switch (type) {
+        case TypedArrayType::Int8:
+            value = getValueFromBuffer<int8_t>(state, byteindex, true);
```
Comment (rerobika) [inline src/runtime/ArrayBufferObject.h]: I think you can directly return in all cases.

Comment (Achie72) [reply src/runtime/ArrayBufferObject.h]: That's a nice catch, that flew past by me. I updated the PR with this change.

Comment (rerobika) [reply src/runtime/ArrayBufferObject.h]: I think this code path has not been updated.

File: src/runtime/ArrayBufferObject.h [hunk 3]
```diff
@@ -98,6 +110,34 @@ class ArrayBufferObject : public Object {
         return true;
     }
 
+    Value getTypedValueFromBuffer(ExecutionState& state, unsigned byteindex, TypedArrayType type)
+    {
+        Value value;
```
Comment (rerobika) [inline src/runtime/ArrayBufferObject.h]: It seems to me that this variable is unused.

**Comment by rerobika** [comment]
@Achie72 please rebase this PR.

**Pull request merged**